### PR TITLE
Keyframes pipeline

### PR DESCRIPTION
**Pull request by adamsvystun**
Implements proper keyframe pipeline.

### CODE REVIEW

**Comment by adamsvystun** [comment]
Dockerfile changes from [https://github.com/BVLC/caffe/blob/master/docker/gpu/Dockerfile](https://github.com/BVLC/caffe/blob/master/docker/gpu/Dockerfile)

File: dockerfile [hunk 1]
```diff
@@ -15,6 +20,25 @@ WORKDIR /comixify
 COPY . /comixify
 RUN python3.6 -m pip install -r requirements.txt
 
+ENV CAFFE_ROOT=/opt/caffe
+WORKDIR $CAFFE_ROOT
+
+ENV CLONE_TAG=1.0
+
+RUN git clone -b ${CLONE_TAG} --depth 1 https://github.com/BVLC/caffe.git . && \
+    pip install --upgrade pip && \
```
Comment (maciej3031) [inline dockerfile]: Not neccessary, pip is already upgraded in line 8?
moreover try to use python3.6 instead 3.

File: dockerfile [hunk 2]
```diff
@@ -15,6 +20,25 @@ WORKDIR /comixify
...
+    cd python && for req in $(cat requirements.txt) pydot; do pip install $req; done && cd .. && \
```
Comment (maciej3031) [inline dockerfile]: again requirements are already installed, line 16. Any reason why to use it again?

File: keyframes/keyframes.py [hunk 3]
```diff
@@ -1,47 +1,154 @@
...
         return chosen_frames
 
     @staticmethod
     def _get_all_frames(video):
         all_frames_tmp_dir = uuid.uuid4()
         os.mkdir(jj(f"{settings.TMP_DIR}", f"{all_frames_tmp_dir}"))
-        call(["ffmpeg", "-skip_frame", "nokey", "-i", f"{video.file.path}", "-vsync", "0", "-qscale:v", "1",
-              "-f", "image2", jj(f"{settings.TMP_DIR}", f"{all_frames_tmp_dir}", "%06d.jpeg")])
-
+        call(["ffmpeg", "-i", f"{video.file.path}", "-vf", "select=not(mod(n\\,15))", "-vsync", "vfr", "-q:v", "2",
```
Comment (maciej3031) [inline keyframes/keyframes.py]: we still need it after adding new key extraction method?

File: keyframes_rl/main.py [hunk 4]
```diff
@@ -0,0 +1,208 @@
+from __future__ import print_function
```
Comment (maciej3031) [inline keyframes_rl/main.py]: I think we do not need this file.

File: keyframes_rl/vsum_tools.py [hunk 5]
```diff
@@ -0,0 +1,113 @@
+import numpy as np
```
Comment (maciej3031) [inline keyframes_rl/vsum_tools.py]: Also this file is not needed I think

File: dockerfile (hunk 2 above)
Comment (adamsvystun) [reply dockerfile]: These are requirements for python layer of caffe, and not the project requirements. They are necessary.

File: keyframes/keyframes.py (hunk 3 above)
Comment (adamsvystun) [reply keyframes/keyframes.py]: We first need to downsample video to frames, then out of those frames select keyframes. It is probably possible to do this through opencv (which I think still uses ffmpeg to process videos) fully in python.

File: keyframes_rl/models.py [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+        if cell == 'lstm':
+            self.rnn = nn.LSTM(in_dim, hid_dim, num_layers=num_layers, bidirectional=True, batch_first=True)
+        else:
+            self.rnn = nn.GRU(in_dim, hid_dim, num_layers=num_layers, bidirectional=True, batch_first=True)
+        self.fc = nn.Linear(hid_dim*2, 1)
+
+    def forward(self, x):
+        h, _ = self.rnn(x)
+        p = F.sigmoid(self.fc(h))
+        return p
```
[the remark this answers is not in the document]
Comment (adamsvystun) [reply keyframes_rl/models.py]: Added.

File: keyframes/tests.py [hunk 7]
```diff
@@ -1,3 +1,56 @@
...
+class KeyframesTestCase(TestCase):
+
+    def setUp(self):
+        f = open(VIDEO_PATH, 'rb')
+        self.video = Video.objects.create(file=File(f))
+
+    def tearDown(self):
+        shutil.rmtree(jj(f"{settings.TMP_DIR}", f"{self.all_frames_tmp_dir}"))
+
+    def test_keyframes(self):
```
[the remark this answers is not in the document]
Comment (adamsvystun) [reply keyframes/tests.py]: I know, ideally, each function should be tested separately, on separate data. But I wanted to check for now if it works, and the problem is that the functions are correlated. I can split testing of each function in separate functions, and then write my own `run()` function, which will run the whole procedure for each testing function. Not sure if it makes any difference then? Or am I not seeing a correct approach?

[the remark this answers is not in the document]
Comment (maciej3031) [reply keyframes/tests.py]: So let's stay with one test for now

**Pull request merged**